clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/provider/agent/GoogleServerGroupCachingAgent.groovy

```diff
@@ -55,70 +61,196 @@ class GoogleServerGroupCachingAgent extends AbstractGoogleCachingAgent {
 
     serverGroups
   }
 
-  CacheResult buildCacheResult(ProviderCache _, List<GoogleServerGroup2> serverGroups) {
+  @Override
+  boolean handles(OnDemandAgent.OnDemandType type, String cloudProvider) {
+    type == OnDemandAgent.OnDemandType.ServerGroup && cloudProvider == GoogleCloudProvider.GCE
+  }
+
+  @Override
+  OnDemandAgent.OnDemandResult handle(ProviderCache providerCache, Map<String, ? extends Object> data) {
+    if (!data.containsKey("serverGroupName") ||
+        data.account != accountName ||
+        data.region != region ||
+        data.zone.empty) {
+      return null
+    }
+
+    GoogleServerGroup2 serverGroup = metricsSupport.readData {
+      getServerGroup(data.serverGroupName as String, data.zone as String)
+    }
+
+    def cacheResultBuilder = new CacheResultBuilder(startTime: Long.MAX_VALUE)
+    CacheResult result = metricsSupport.transformData {
+      buildCacheResult(cacheResultBuilder, [serverGroup])
+    }
+
+    if (result.cacheResults.values().flatten().empty) {
+      // Avoid writing an empty onDemand cache record (instead delete any that may have previously existed).
+      providerCache.evictDeletedItems(ON_DEMAND.ns, [Keys.getServerGroupKey(serverGroup.name,
+                                                                            accountName,
+                                                                            serverGroup.zone)])
+    } else {
+      metricsSupport.onDemandStore {
+        def cacheData = new DefaultCacheData(
+            Keys.getServerGroupKey(data.serverGroupName as String, accountName, data.zone as String),
+            10 * 60, // ttl is 10 minutes
+            [
+                cacheTime     : System.currentTimeMillis(),
+                cacheResults  : objectMapper.writeValueAsString(result.cacheResults),
+                processedCount: 0,
+                processedTime : null
+            ],
+            [:]
+        )
+
+        providerCache.putCacheData(ON_DEMAND.ns, cacheData)
+      }
+    }
+
+    Map<String, Collection<String>> evictions = [:]
+    if (!serverGroup) {
+      evictions[SERVER_GROUPS.ns].add(Keys.getServerGroupKey(data.serverGroupName as String,
+                                                             accountName,
+                                                             data.zone as String))
+    }
+
+    log.info("On demand cache refresh succeeded. Data: ${data}")
+
+    return new OnDemandAgent.OnDemandResult(
+        sourceAgentType: getOnDemandAgentType(),
+        cacheResult: result,
+        evictions: evictions
+    )
+  }
+
+  @Override
+  Collection<Map> pendingOnDemandRequests(ProviderCache providerCache) {
+    def keyOwnedByThisAgent = { Map<String, String> parsedKey ->
+      parsedKey && parsedKey.account == accountName && region == credentials.regionFromZone(parsedKey.zone)
+    }
+
+    def keys = providerCache.getIdentifiers(ON_DEMAND.ns).findResults { String key ->
```

Comment:
`.findAll { key -> keyOwnedByThisAgent(Keys.parse(key)) }` might be neater here.